src/responder/kcm/secrets/secrets.c

```diff
@@ -181,11 +188,166 @@ static struct ldb_dn *per_uid_container(TALLOC_CTX *mem_ctx,
     return uid_base_dn;
 }
 
+static errno_t get_secret_expiration_time(uint8_t *key, size_t key_length,
+                                          uint8_t *sec, size_t sec_length,
+                                          time_t *_expiration)
+{
+    errno_t ret;
+    TALLOC_CTX *tmp_ctx;
+    time_t expiration = 0;
+    struct cli_creds client = {};
+    struct kcm_ccache *cc;
+    struct sss_iobuf *iobuf;
+    krb5_creds **cred_list, **cred;
+    const char *key_str;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    key_str = talloc_strndup(tmp_ctx, (const char *) key, key_length);
+    if (key_str == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    iobuf = sss_iobuf_init_readonly(tmp_ctx, sec, sec_length);
+    if (iobuf == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = sec_kv_to_ccache_binary(tmp_ctx, key_str, iobuf, &client, &cc);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    cred_list = kcm_cc_unmarshal(tmp_ctx, NULL, cc);
+    if (cred_list == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    for (cred = cred_list; *cred != NULL; cred++) {
+        if ((*cred)->times.endtime != 0) {
```

Comment:
There is no order in the list.

it is not possible to order the results by the expiration time because it isn't an LDAP attribute but a kerberos field stored together with many others in the `secret` LDAP attribute that must be decoded one by one.

If it was a separate attribute, we could use the sorting controls (assuming LDB implements them) to get the results sorted. However, I'm not sure that is a better solution considering we are not talking to a real LDAP server but to a local library. In that case, the library will sort the results on the same host (same memory, same CPUs, etc.) which is more expensive that just looking for the lowest element (O(n)).

Reply:
OK, thanks for the explanation -  but there is a think that I do not understand.

You find first credentials in the list with `times.endtime` set. You save the value into `expiration` variable. Then you continue the cycle and to the same without any further check of that value. So as the result - we pick always the last item from the list with times.endtime set. Why not the first one?

    for (cred = cred_list; *cred != NULL; cred++) {
        if ((*cred)->times.endtime != 0) {
            expiration = (time_t) (*cred)->times.endtime;
            break; // we found something
        }
    }

or bigest/smallest one

    for (cred = cred_list; *cred != NULL; cred++) {
        if ((*cred)->times.endtime != 0) {
            // some check that newly found time is more interesting for the purpose?
            if (expiration >=<? (time_t) (*cred)->times.endtime) {
                expiration = (time_t) (*cred)->times.endtime;
            }
        }
    }

if you see what I mean...

Reply:
The code is supposed to look for the oldest expiration time, which is the smallest one. But that condition is missing. You found a real bug.

And all this worked in my tests because the oldest one happens to be the first.

Reply:
I see what you mean. As you can see in my first answer, I though you were talking about the different secrets, but it was not the case. You are talking about each credential in a single secret. And since in that case we know there is a single expiration time, it is a good thing to leave the loop as soon as we find it. I was convinced I had done it.

It's fixed now.